flocker/acceptance/testtools.py

```diff
@@ -324,32 +325,66 @@ def get_nodes(test_case, num_nodes):
-                          u"applications": {}}
     getting = get_test_cluster(reactor)
 
-    def got_cluster(cluster):
-        # Remove all existing containers; we make sure to pass in node
-        # hostnames since we still rely on flocker-deploy to distribute SSH
-        # keys for now.
-        clean_deploy = {u"version": 1,
-                        u"nodes": {node.address: [] for node in cluster.nodes}}
-        flocker_deploy(test_case, clean_deploy, clean_applications)
-        return cluster
-    getting.addCallback(got_cluster)
-
-    def no_containers(cluster):
-        d = cluster.current_containers()
-        d.addCallback(lambda result: len(result[1]) == 0)
-        return d
-    getting.addCallback(lambda cluster:
-                        loop_until(lambda: no_containers(cluster)))
-
-    def clean_zfs(_):
-        for node in reachable_nodes:
-            _clean_node(test_case, node)
-    if get_volume_backend(test_case) == b'zfs':
-        getting.addCallback(clean_zfs)
+    def clean_containers(cluster):
+        get_containers = cluster.configured_containers()
+
+        def delete_containers(result):
+            cluster, containers = result
+            results_list = []
+            for container in containers:
+                results_list.append(
+                    cluster.remove_container(container[u"name"]))
+            deleting = gatherResults(results_list)
+            deleting.addCallback(lambda _: cluster)
+            return deleting
+
+        get_containers.addCallback(delete_containers)
+
+        def no_containers(cluster):
+            d = cluster.current_containers()
+            d.addCallback(lambda result: len(result[1]) == 0)
+            return d
+
+        def check_clean(cluster):
+            d = loop_until(lambda: no_containers(cluster))
+            d.addCallback(lambda _: cluster)
+            return d
+
+        get_containers.addCallback(check_clean)
+        return get_containers
+
+    getting.addCallback(clean_containers)
+
+    def clean_datasets(cluster):
+        get_datasets = cluster.configured_datasets()
+
+        def delete_datasets(datasets):
+            results_list = []
+            for dataset in datasets:
+                results_list.append(
+                    cluster.delete_dataset(dataset[u"dataset_id"]))
+            deleting = gatherResults(results_list)
+            deleting.addCallback(lambda _: cluster)
+            return deleting
+
+        get_datasets.addCallback(delete_datasets)
+
+        def no_datasets(cluster):
+            d = cluster.datasets_state()
+            d.addCallback(lambda datasets: len(datasets) == 0)
+            return d
+
+        def check_datasets(cluster):
+            d = loop_until(lambda: no_datasets(cluster))
+            d.addCallback(lambda _: cluster)
+            return d
+
+        get_datasets.addCallback(check_datasets)
+        return get_datasets
+
+    getting.addCallback(clean_datasets)
```

Comment:
Same pattern repeated here for datasets and containers.  This could be avoided with a little helper that takes `cluster.configured_{datasets,containers}`, `cluster.{delete_dataset,remove_container}`, `cluster.{dataset_state,current_containers}` (particularly if the weird tuple return type of `current_containers` gets fixed).

This probably makes sense as a top-level function in the module, too, rather than a nested helper.